subsys/lw_sched/lw_scheduler.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright(c) 2023 Intel Corporation
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/lw_sched/lw_sched.h>
+
+extern void lw_task_insert(struct lw_scheduler *sched, struct lw_task *task);
+
+static void timer_expiry(struct k_timer *timer)
+{
+	struct lw_scheduler *sched;
+
+	/* Get the scheduler based on the timer */
+
+	sched = CONTAINER_OF(timer, struct lw_scheduler, timer);
+
+	k_sem_give(&sched->sem);   /* Wake the scheduler's thread */
+}
+
+static void lw_scheduler_reorder(struct lw_scheduler *sched)
+{
+	sys_dnode_t *reorder;
+	struct lw_task *task;
+
+	/*
+	 * Loop through each node in the reorder list. Remove the associated
+	 * task from the list of tasks and insert it at the correct priority.
+	 */
+
+	while ((reorder = sys_dlist_get(&sched->reorder)) != NULL) {
```

Comment:
This makes insertion/removal O(N), which makes complicated priority management expensive for large task lists.  Given that the whole point to this is to be "light weight" and suitable for using with hundreds/thousands of tasks, that's probably something we want to fix.

Use a rbtree (or some other log-or-better-time collection, but we already have an rbtree) instead?  Doesn't look like it should be difficult to port.

Reply:
Working on it.

Reply:
Now I remember why I went with the doubly linked list--despite the the O(N) cost ... removing nodes from an RB tree while it is being traversed can lead to nodes being visited twice or skipped, so I went with simplicity.